packages/adapter-fauna/src/index.ts

```diff
@@ -0,0 +1,168 @@
+import { LuciaError, generateRandomString } from "lucia-auth";
+import type { Adapter } from "lucia-auth";
+import faunadb from "faunadb";
+import { convertUserResponse } from "./utils.js";
+
+const { query, Client, errors } = faunadb;
+const { FaunaError } = errors;
+const q = query;
+
+const adapter = (// @ts-ignore
+	faunaClient: Client, // @ts-ignore
+	errorHandler: (error: FaunaError) => void = () => {
+	}): Adapter => {
+	return {
+		getUser: async (userId) => {
+			try {
+				const response = await faunaClient.query(q.Map(q.Paginate(q.Match(q.Index("user_by_id"), userId)), q.Lambda("x", q.Get(q.Var("x")))));
+				const users = response.data;
+				return users.length !== 0 ? users[0].data : null;
+
+			} catch (e) {
+				errorHandler(e as any);
+				throw e;
+			}
+		}, getUserByProviderId: async (providerId) => {
+			try {
+				const response = await faunaClient.query(q.Map(q.Paginate(q.Match(q.Index("user_by_providerid"), providerId)), q.Lambda("x", q.Get(q.Var("x")))));
+				const users = response.data;
+				return users.length !== 0 ? users[0].data : null;
+
+			} catch (e) {
+				errorHandler(e as any);
+				throw e;
+			}
+		}, getSessionAndUserBySessionId: async (sessionId) => {
+			try {
+				const sessionsResponse = await faunaClient.query(q.Map(q.Paginate(q.Match(q.Index("session_by_id"), sessionId)), q.Lambda("x", q.Get(q.Var("x")))));
+				const sessions = sessionsResponse.data;
+				const session = sessions.length !== 0 ? sessions[0].data : null;
+				if (!session) return null;
+
+				const userResponse = await faunaClient.query(q.Map(q.Paginate(q.Match(q.Index("user_by_id"), session["user_id"])), q.Lambda("x", q.Get(q.Var("x")))));
+				const users = userResponse.data;
+				const user = users.length !== 0 ? users[0].data : null;
+				if (!user) return null;
+
+				return {
+					user: user, session: session
+				};
+			} catch (e) {
+				errorHandler(e as any);
+				throw e;
+			}
+		}, getSession: async (sessionId) => {
+			try {
+				const response = await faunaClient.query(q.Map(q.Paginate(q.Match(q.Index("session_by_id"), sessionId)), q.Lambda("x", q.Get(q.Var("x")))));
+				const sessions = response.data;
+				return sessions.length !== 0 ? sessions[0].data : null;
+
+			} catch (e) {
+				errorHandler(e as any);
+				throw e;
+			}
+		}, getSessionsByUserId: async (userId) => {
+			try {
+				const response = await faunaClient.query(q.Map(q.Paginate(q.Match(q.Index("session_by_userid"), userId)), q.Lambda("x", q.Get(q.Var("x")))));
+				return convertUserResponse(response.data);
+
+			} catch (e) {
+				errorHandler(e as any);
+				throw e;
+			}
+		}, setUser: async (userId, userData) => {
+			try {
+				const response = await faunaClient.query(q.Create(q.Collection("users"), {
+					data: {
+						id: userId || generateRandomString(20),
```

Comment:
Just for keep consistency between adapters, can you throw a custom error when the `userId` is `null` and mention that the adapter doesn't generate its own id? (Remember you remove the import statement)
```ts
if (userId === null) throw new Error("auto user id generation is not supported");
```

What the "correct" behavior should be can be discussed later.

Reply:
How about adding a new LuciaError for this?